lib/fathead/diveintopython/parse.py

```diff
@@ -0,0 +1,210 @@
+        for header in self.soup.find_all('h2'):
+            section = self.parse_section(header)
+            self.main_sections.append(section)
+
+        for header in self.soup.find_all('h3'):
+            section = self.parse_section(header)
+            section['example'] = self.get_example(header)
+            section['abstract'] = '{}<br><pre><code>{}</code></pre>'.format(section['paragraph'], section['example'])
+            self.sub_sections.append(section)
+
+        self.parsed_data = self.main_sections + self.sub_sections
+
+    def get_data(self):
+        """
+        Get the parsed data.
+        Returns:
+            self.parsed_data: Dict containing necessary data elements
+        """
+        return self.parsed_data
+
+    def get_example(self, soup):
+        """
+        Gets the next code example after soup tag object
+        Args:
+            Soup tag object
+        Returns:
+            Text inside <pre> tag
+        """
+        return soup.findNext('pre').get_text().replace('\n', '\\n')
+
+    def get_url(self, soup):
+        """
+        Parses url to the content
+        Args:
+            Soup tag object
+        Returns:
+            Url with anchor
+        """
+
+        anchor = soup.get('id')
+
+        if anchor:
+            parsed_url = '{}#{}'.format(os.path.join(URL_ROOT, self.file_name), anchor)
+        else:
+            parsed_url = os.path.join(URL_ROOT, file_name)
+
+        return parsed_url
+
+    def parse_section(self, soup):
+        """
+        Parses main and sub sections in soup object.
+
+        Args:
+            soup: Soup tag object
+        Returns:
+            Section data, containing title, first paragraph and url to content.
+        """
+        first_paragraph = self.clean_formatting(soup.findNext('p').get_text())
+
+        return {
+            'title': self.clean_formatting(soup.get_text()),
+            'paragraph': first_paragraph,
+            'anchor': self.get_url(soup),
+            'url': os.path.join(URL_ROOT, self.file_name)
+        }
+
+    def clean_formatting(self, text):
+        """
+        Fixes up some weird double spacing and newlines.
+        Args:
+            text: Piece of text to be fixed.
+
+        Returns:
+            Given text without double spacing and new lines.
+        """
+        text = text.replace('  ', ' ').replace('\n', ' ').replace('\\n', r'\\n')
+        return text.strip()
+
+
+class PythonDataOutput(object):
+    """
+    Object responsible for outputting data into the output.txt file
+    """
+    def __init__(self, data):
+        self.data = data
+
+    def create_file(self):
+        """
+        Iterate through the data and create the needed output.txt file,
+        appending to file as necessary.
+        """
+        with open('output.txt', 'w') as output_file:
+            for data_element in self.data:
+                title = data_element.get('title')
+                abstract = data_element.get('abstract')
```

Comment:
Could simplify this to:
```
first_paragraph = data_element.get('first_paragraph')
abstract = data_element.get('abstract') or first_paragraph
```
to avoid later-on checking for None.